developer/src/kmc-copy/src/KeymanProjectCopier.ts

```diff
@@ -194,61 +200,64 @@ export class KeymanProjectCopier implements KeymanCompiler {
    * @param source
    * @returns a promise: GitHub reference to the source for the keyboard, or null on failure
    */
   private async getGitHubSourceProject(source: string): Promise<GitHubRef> {
-    const parts = source.split(':');
-    if(parts.length < 3 || parts.length > 4 || !parts[1].match(/^[a-z0-9-]+\/[a-z0-9._-]+$/i)) {
-      // https://stackoverflow.com/questions/59081778/rules-for-special-characters-in-github-repository-name
-      this.callbacks.reportMessage(CopierMessages.Error_InvalidGitHubSource({source}));
-      return null;
+    const parts: GitHubUrls.GitHubRegexMatchArray =
+      GitHubUrls.GITHUB_URI_OPTIONAL_PROTOCOL.exec(source) ??
+      GitHubUrls.GITHUB_RAW_URI.exec(source);
+    if(!parts) {
+      throw new Error('Expected GITHUB_URI_OPTIONAL_PROTOCOL or GITHUB_RAW_URI to match');
     }
 
-    const origin = parts[1].split('/');
+    const ref: GitHubRef = new GitHubRef(parts);
 
-    const ref: GitHubRef = new GitHubRef({
-      owner: origin[0],
-      repo: origin[1],
-      branch: null,
-      path: null
-    });
-
-    if(parts.length == 4) {
-      ref.branch = parts[2];
-      ref.path = parts[3];
-    } else {
+    if(!ref.branch) {
       ref.branch = await this.cloudSource.getDefaultBranchFromGitHub(ref);
       if(!ref.branch) {
         this.callbacks.reportMessage(CopierMessages.Error_CouldNotFindDefaultBranchOnGitHub({ref: ref.toString()}));
         return null;
       }
-      ref.path = parts[2];
-
     }
+    if(!ref.path) {
+      ref.path = '/'
+    };
     if(!ref.path.startsWith('/')) {
       ref.path = '/' + ref.path;
     }
 
+    if(ref.path != '/') {
+      if(!ref.path.endsWith('.kpj')) {
+        // Assumption, project filename matches folder name
+        if(ref.path.endsWith('/')) {
+          ref.path = ref.path.substring(0, ref.path.length-1);
+        }
+        ref.path = ref.path + '/' + this.callbacks.path.basename(ref.path) + '.kpj';
+      }
+    }
+
     return ref;
   }
 
   /**
    * Resolve path to Keyman Cloud source (which is on GitHub), which must be in
    * the following format:
-   *   `cloud:keyboard_id|model_id`
+   *   `cloud:keyboard_id`, or
+   *   `cloud:model_id`, or
+   *   `https://keyman.com/keyboards/keyboard_id`
    * The `keyboard_id` parameter should be a valid id (a-z0-9_), as found at
-   * https://keyman.com/keyboards; alternativel if it is a model_id, it should
+   * https://keyman.com/keyboards; alternatively if it is a model_id, it should
    * have the format author.bcp47.uniq
    * @param source
    * @returns a promise: GitHub reference to the source for the keyboard, or null on failure
    */
   private async getCloudSourceProject(source: string): Promise<GitHubRef> {
-    const parts = source.split(':');
-    const id = parts[1];
+    const parts = CLOUD_URI.exec(source) ?? KEYMANCOM_CLOUD_URI.exec(source);
+    const id: string = parts.groups.id;
 
     const isModel = /^[^.]+\.[^.]+\.[^.]+$/.test(id);
```

Comment:
Would one of the regexes work?
https://github.com/keymanapp/keyman/blob/80c9cfb5b92743b7c6ffc7ba260cf1ca483841ed/developer/src/common/web/utils/src/source-filename-patterns.ts#L30

Reply:
no, because those regexes all have `.model.` in them. We could build a shared regex I guess but I am not really keen on more refactoring for it.